DDF-1601 Implement dynamic port allocation

harrison-tarr opened the pull request:
Port numbers and URLs based thereon were declared as finals in
AbstractIntegrationTest. These are now dynamically allocated
based on what ports are in use. Test files had to be changed to
load in the new port numbers instead of using the class defaults.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf/287)
<!-- Reviewable:end -->

harrison-tarr commented:
To hero: run 2 PRBS at the same time.

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -82,33 +83,38 @@
     protected static final String KARAF_VERSION = "2.4.3";
 
     // TODO: Use the Camel AvailablePortFinder.getNextAvailable() test method
```
Is this comment OBE?

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -166,11 +172,33 @@ public void initFacades() {
     public Option[] config() throws URISyntaxException {
+        basePort = findPortNumber(20000);
+
+        HTTP_PORT = String.valueOf(basePort + 1);
```
Instead of keeping/modifying all these constant strings, how about using a `final static Map` with these constants changed to keys? I think you could end up managing fewer moving pieces that way. Maybe.

lessarderic replied:
Could also create a static inner class (e.g., `Ports`) that calls `findPortNumber` in its ctor and has getters for the different port values.

lessarderic commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -166,11 +172,33 @@ public void initFacades() {
+        SSH_PORT = String.valueOf(basePort + 3);
+        RMI_SERVER_PORT = String.valueOf(basePort + 4);
+        RMI_REG_PORT = String.valueOf(basePort + 5);
+        LOGGER.debug(HTTP_PORT + " " + HTTPS_PORT + " " + SSH_PORT + " " + RMI_SERVER_PORT + " "
```
Adding labels to the log would be helpful. Otherwise it's just going to be a bunch of numbers with no context.

brendan-hofmann commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 192 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f-KIeFjGkx07Pp_Dq)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L192)):</sup>
Could we replace this method with `AvailablePortFinder.getNextAvailable(baseport?)`?
<!-- Reviewable comment -K1f-KIeFjGkx07Pp_Dr:-184302392 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java, line 92 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f00rpcjHRkHdQtsSO)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java#L92)):</sup>
So we have to remember to call this in every `@BeforeExam`? :crying_cat_face: 
<!-- Reviewable comment -K1f00rpcjHRkHdQtsSP:-1472874418 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestFederation.java, line 104 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f0S9duDf1XBkTVUbD)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestFederation.java#L104)):</sup>
Why not use the roots in AbstractIntegrationTest instead of duplicating those values here?
<!-- Reviewable comment -K1f0S9duDf1XBkTVUbE:1832182801 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 214 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f0b6P5dCz4c1T8AnH)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L214)):</sup>
Be nice to tidy up all these string literals with the appropriate constants some day...
<!-- Reviewable comment -K1f0b6P5dCz4c1T8AnI:789649855 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

lessarderic commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java:
```diff
@@ -89,10 +89,11 @@ public static String ingest(String data, String mimeType) {
         waitForAllBundles();
         waitForCatalogProvider();
-        waitForHttpEndpoint(SERVICE_ROOT + "/catalog/query");
+        waitForHttpEndpoint(AbstractIntegrationTest.SERVICE_ROOT + "/catalog/query");
```
Any reason we have to qualify with `AbstractIntegrationTest`?

harrison-tarr commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 192 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f-KIeFjGkx07Pp_Dq)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L192)):</sup>
No because `AvailablePortFinder.getNextAvailable(baseport?)` closes the socket and port and also does not jump by 10. Other than that, this implementation is basically the same.
<!-- Reviewable comment -K1fU0o50yceqJJqvLdh:726142928 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java, line 96 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdCkeQzFdUli-Fm-r1-96)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java#L96)):</sup>
I just forgot to take it out. Had put it in to see if qualifying would make a difference in the value. (It didn't)
<!-- Reviewable comment -K1fUNc6Lqux4N42iS3s:2057250944 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
Review status: 0 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 177 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r1-177)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L177)):</sup>
fixed
<!-- Reviewable comment -K1fXkT6ksO_OgW6EPzI:-467915377 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -81,38 +82,109 @@
+        String getUrl(Integer basePort) {
+            String returnUrl = "";
+            for (int i = 0; i < pieces.length; i++) {
+                if (null != pieces[i] && pieces[i] instanceof Port) {
```
You can skip the null test if the next test is `instanceof`.

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -81,38 +82,109 @@
+            this.root = base;
+        }
+
+        String getUrl(Integer basePort) {
```
I think you could probably write this more concisely with a `String.format`.

coyotesqrl replied:
If you change your `Enum[]` to a `List<Enum>`, you should be able to replace this method's guts with...
```
            return pieces.stream()
                    .map(p -> {
                        if (p instanceof Port) {
                            return ((Port) p).getPort(basePort);
                        } else if (p instanceof Url) {
                            return ((Url) p).getUrl(basePort);
                        }
                        return "";
                    })
                    .reduce("", (acc, val) -> acc + val) + root;
```

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -81,38 +82,109 @@
+            return returnUrl;
+        }
+
+        String getUrl() {
```
This method's impl looks identical to your existing implementation of the overrided version. You could just call the other method, passing in the `basePort` field as a parameter.

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -276,20 +364,24 @@ protected UrlResourceReaderConfigurator getUrlResourceReaderConfigurator() {
-                replaceConfigurationFile("etc/ddf.security.sts.client.configuration.cfg", new File(
-                        this.getClass().getResource("/ddf.security.sts.client.configuration.cfg")
-                                .toURI())), replaceConfigurationFile(
+                editConfigurationFilePut("etc/ddf.security.sts.client.configuration.cfg", "address",
```
Why did this change from a full replace of the sts client config file to an edit?

harrison-tarr replied:
it loaded in a hard-coded string directly from the file that had `9993` as a port in it; this way we put in the correct port and avoid that

coyotesqrl commented:
I'm concerned about removing those protected statics now that I'm thinking about the ramifications for existing subclasses.

brendan-hofmann commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 97 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f_k6_n8qe-sIPA7wg)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L97)):</sup>
The naming conventions in here are a bit confusing. For example, enums are typically all caps, due to their inherently static nature. I'd like to see more specific names as well, otherwise it can make the usage hard to read. Maybe something like `DYNAMIC_PORT` and `DYNAMIC_URL`?
<!-- Reviewable comment -K1f_k6_n8qe-sIPA7wh:602954503 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 177 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r1-177)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L177)):</sup>
Not quite what I had in mind. See my comments above.
<!-- Reviewable comment -K1jHgM9-aCZdMJdhE1M:1551715465 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 97 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f_k6_n8qe-sIPA7wg)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L97)):</sup>
The enum name `Port` is fine. The enum values need to be all caps.

Seems to me like the enum approach is a little confusing and cumbersome. Couldn't we just create a simple `Ports` class, instantiate it in the `config()` method and use that instance in the tests instead of using the constants? I know it'll require some refactoring of the existing tests but I think it'll keep the code cleaner and more maintainable.

```
private static class Ports {
  public static class Port {
    private final int portNumber;
    private final String systemPropertyName;

    private Port(int portNumber, String systemPropertyName) {
      this.portNumber = portNumber;
      this.systemPropertyName = systemPropertyName;
    }
    public int getPortNumber() {
      return portNumber;
    }
    public String getSystemPropertyName() {
      return systemPropertyName;
    }
    public String getSystemPropertyPortNumber() {
      return System.getProperty(systemPropertyName);
    }
  }
  private static final int HTTP_PORT_OFFSET = 1;
  ...

  private final int basePort;
  private final Port httpPort;
  ...

  public Ports() {
    basePort = findPortNumber(20000);
    httpPort = new Port(basePort + HTTP_PORT_OFFSET, "org.codice.ddf.system.httpPort");
  }

  public Port getHttpPort() {
    return httpPort;
  }
}
```
<!-- Reviewable comment -K1jDN4-Rseg_7hnxXsA:-1851520453 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 124 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1jGJ0CtvdocAGOgNaa)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L124)):</sup>
Do we really need this method? If so, the name is a little misleading.
<!-- Reviewable comment -K1jGJ0DNWLnl7KA6qiB:1435382602 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 131 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1jDrzb65nxrVPRIi5r)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L131)):</sup>
Seems weird to me to put `SECURE_ROOT` and `INSECURE_ROOT` as enum values as they are incomplete URLs. Can't we just use plain constants for those instead?
<!-- Reviewable comment -K1jDrzb65nxrVPRIi5s:278490189 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 147 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1jH2dOHo_07Hphbsxn)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L147)):</sup>
Instead of having different ctors and "complex" logic in `getUrl()`, can't we just have a single ctor that takes in String..., concatenate them to build the URL and save it off?
<!-- Reviewable comment -K1jH2dPlKSiSvqLpE3G:8731623 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

AzGoalie commented:
<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 100 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1jO5-L7_yfXdBwSXzQ)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L100)):</sup>
the https port should be 'org.codice.ddf.system.httpsPort'
<!-- Reviewable comment -K1jO5-L7_yfXdBwSXzR:1203499138 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

clockard commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -81,38 +82,113 @@
 
-    protected static final String SERVICE_ROOT = SECURE_ROOT + HTTPS_PORT + "/services";
+    public enum DynamicPort {
+        BASE_PORT("org.codice.ddf.system.basePort"), HTTP_PORT(
```
Think this should be "org.codice.ddf.system.port"

harrison-tarr replied:
This is a system property just used for tests. The baseport is used to define an offset that the test can then use to assign its ports without interfering with other tests.

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -81,38 +82,121 @@
+    }
 
-    protected static final String CSW_SOURCE_ID = "cswSource";
+    public static class DynamicString {
```
Since this is used for URLs exclusively, it would probably be good to give it a less generic name. Something like...`UrlBuilder` or `DynamicUrl` or some such.

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -81,38 +82,121 @@
+            this.root = base;
+        }
+
+//        public DynamicString(String piece, String base) {
```
Commented-out methods should be removed.

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -81,38 +82,121 @@
-    protected static String ddfHome;
+        private final List<Object> pieces = new ArrayList<>();
+
+        public DynamicString(String base) {
```
You might consider looking at the work done on the new [`WaitCondition`](https://github.com/codice/ddf/blob/master/distribution/test/itests/test-itests-common/src/test/java/ddf/common/test/WaitCondition.java) class. I think a fluent builder would read a bit cleaner for this use. e.g.
```
UrlBuilder
  .from(base)
  .with(piece)
  .with(piece)
  .toUrl();
```
You could override the `with` method to take either a `DynamicPort` or another instance of `this` and could also add a `withBasePort` method to support that functionality.

It's certainly not a necessary change but I think it might lead to slightly cleaner uses. Worth a quick test, if nothing else.

lessarderic commented:
Review status: 0 of 7 files reviewed at latest revision, 25 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 125 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r5-125)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L125)):</sup>
:+1: 
<!-- Reviewable comment -K288crjzL-12smyRBSw:-510466975 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 133 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r5-133)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L133)):</sup>
You can also look at https://docs.oracle.com/javaee/6/api/javax/ws/rs/core/UriBuilder.html to see if you can use it somehow.
<!-- Reviewable comment -K2895PGPHrXOBDTXV5x:-1733382740 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 168 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28AuZoiT1W7rkzBHwh)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L168)):</sup>
Can't we avoid using an untyped `ArrayList<Object>` and all those `instanceof` and instead have typed fields for the different URL pieces, e.g., `String baseUrl`, `int port`, `ArrayList<String> contextPath`?
<!-- Reviewable comment -K28AuZpf3LzJ-wnKlN0:-1643731805 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 361 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28CEdpBVkhP6blVYyM)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L361)):</sup>
We should be able to remove all those `DynamicPort.` prefixes.
<!-- Reviewable comment -K28CEdqp7BfK_DebgdH:2093143346 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 215 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28CjbawWfXlWWxndi5)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L215)):</sup>
Why not create a constant using `DynamicUrl` instead?
<!-- Reviewable comment -K28CjbawWfXlWWxndi6:-369175506 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 244 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28Clq_WzdHj3MG-vqE)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L244)):</sup>
Could re-use the constant created above :-)
<!-- Reviewable comment -K28Clq_WzdHj3MG-vqF:-1189938723 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 523 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28D8CJw5demqLA9rPZ)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L523)):</sup>
Here too.
<!-- Reviewable comment -K28D8CKGF8zu28VUUoT:-1213865979 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 568 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28DDBIeAaGMkKmDkqM)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L568)):</sup>
`DynamicUrl` constant.
<!-- Reviewable comment -K28DDBIeAaGMkKmDkqN:-66945697 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

coyotesqrl commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java:
```diff
@@ -81,38 +88,114 @@
+        }
+
+        public String getUrl() {
+            if (null != port) {
```
Couldn't someone (incorrectly) invoke one of the `root`, `port` flavored ctors with a null port? And wouldn't that then mess up the logic here?

Maybe `@NotNull` annotations or `IllegalArgumentException`s?

lessarderic commented:
Reviewed 3 of 6 files at r3, 1 of 3 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 27 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 183 \[r6\]](https://reviewable.io:443/reviews/codice/ddf/287#-K29-_yb0O0gfcdrGLto)** ([raw file](https://github.com/codice/ddf/blob/24d68cd1d5d9d5a723af72f15d86cf15521747d2/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L183)):</sup>
Should be able to static import `DynamicUrl.SECURE_ROOT` and `INSECURE_ROOT` as well.
<!-- Reviewable comment -K29-_ycL1IQ1iBcbD0R:1926515535 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
Review status: 6 of 7 files reviewed at latest revision, 27 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 85 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r1-85)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L85)):</sup>
Done.
<!-- Reviewable comment -K293pyG49MMll7de6EY:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 177 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r1-177)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L177)):</sup>
Done.
<!-- Reviewable comment -K294dxXJRu8ATkVqDqp:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 182 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r1-182)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L182)):</sup>
Removed.
<!-- Reviewable comment -K294eLfhgDhawFm9kLZ:1740281843 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 97 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f_k6_n8qe-sIPA7wg)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L97)):</sup>
Done.
<!-- Reviewable comment -K293ry-npIh1sLrW-zS:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 100 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1jO5-L7_yfXdBwSXzQ)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L100)):</sup>
Done.
<!-- Reviewable comment -K293tCLJD7m_Tjb3WFp:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 124 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1jGJ0CtvdocAGOgNaa)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L124)):</sup>
Removed.
<!-- Reviewable comment -K293uYta36aLmewhDZk:1740281843 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 131 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1jDrzb65nxrVPRIi5r)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L131)):</sup>
Switched to constants in the DynamicUrl class.
<!-- Reviewable comment -K293yXOAWsqVbGN2raM:-1726250326 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 147 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1jH2dOHo_07Hphbsxn)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L147)):</sup>
Refactored.
<!-- Reviewable comment -K2943OLSaYEZbNg2q7Q:2082914472 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 162 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r2-162)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L162)):</sup>
Refactored the DynamicUrl class to avoid this altogether.
<!-- Reviewable comment -K2946zTDhYeQOVOgCFK:119481256 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 165 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r2-165)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L165)):</sup>
Done.
<!-- Reviewable comment -K294A-vkvpsF-mCms1b:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 175 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r2-175)** ([raw file](https://github.com/codice/ddf/blob/e443e0c1de9b3714eb150cb73e0697f98cd30982/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L175)):</sup>
Done.
<!-- Reviewable comment -K294BPa6wJSLgNWUaM8:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 125 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r5-125)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L125)):</sup>
Renamed.
<!-- Reviewable comment -K293wczlSZDYcx_Jd3Q:1755713581 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 142 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r5-142)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L142)):</sup>
Done.
<!-- Reviewable comment -K2944PvDsgsUEj8mRen:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 168 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28AuZoiT1W7rkzBHwh)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L168)):</sup>
Refactored DynamicUrl class
<!-- Reviewable comment -K294HJyCNpyrjaYV-G9:141859522 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 361 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28CEdpBVkhP6blVYyM)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L361)):</sup>
Removed.
<!-- Reviewable comment -K294fdQ3verky6-Qqrt:1740281843 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 163 \[r6\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r6-163)** ([raw file](https://github.com/codice/ddf/blob/24d68cd1d5d9d5a723af72f15d86cf15521747d2/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L163)):</sup>
Added @NotNull annotations in constructors and IllegalArgumentExceptions (just in case)
<!-- Reviewable comment -K294Bx3VEzp1QN7WwYV:-880416571 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 183 \[r6\]](https://reviewable.io:443/reviews/codice/ddf/287#-K29-_yb0O0gfcdrGLto)** ([raw file](https://github.com/codice/ddf/blob/24d68cd1d5d9d5a723af72f15d86cf15521747d2/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L183)):</sup>
Fixed. In some cases, where there is only 1 instance of `DynamicUrl.SECURE_ROOT` or `DynamicPort.HTTPS_PORT`, for example, I left it with the `DynamicPort` prefix to avoid needing the import. I can change this though and always statically import those.
<!-- Reviewable comment -K294IyhY351QalJLbFA:767830448 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java, line 92 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f00rpcjHRkHdQtsSO)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java#L92)):</sup>
Unfortunately.
<!-- Reviewable comment -K294iLIghpxGdItChq_:-1266124911 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestFederation.java, line 104 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f0S9duDf1XBkTVUbD)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestFederation.java#L104)):</sup>
Done.
<!-- Reviewable comment -K294k7kwfO7nmaXTUjz:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 214 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f0b6P5dCz4c1T8AnH)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L214)):</sup>
Done.
<!-- Reviewable comment -K294loCfRE2PF0_KnuW:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 215 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28CjbawWfXlWWxndi5)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L215)):</sup>
Done.
<!-- Reviewable comment -K294mvMxWXkGxKRuGGq:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 244 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28Clq_WzdHj3MG-vqE)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L244)):</sup>
Done.
<!-- Reviewable comment -K294nq2sb9mKBHZErsg:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 523 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28D8CJw5demqLA9rPZ)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L523)):</sup>
Done.
<!-- Reviewable comment -K294oECYukTqvdVrFwg:-499146009 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java, line 568 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K28DDBIeAaGMkKmDkqM)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestSecurity.java#L568)):</sup>
Done.
<!-- Reviewable comment -K294oaN1c4FWYT_4GOc:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
Review status: 6 of 7 files reviewed at latest revision, 27 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 133 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1etSdBcof0eSlrv8zE-r5-133)** ([raw file](https://github.com/codice/ddf/blob/d19634756fe978e430c96327d0aabe5f8a2bf4a1/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L133)):</sup>
Went with a simpler constructor that should remove the need for this.
<!-- Reviewable comment -K294tq7P6diQgqbqdTj:132974958 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 22 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 183 \[r6\]](https://reviewable.io:443/reviews/codice/ddf/287#-K29-_yb0O0gfcdrGLto)** ([raw file](https://github.com/codice/ddf/blob/24d68cd1d5d9d5a723af72f15d86cf15521747d2/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L183)):</sup>
That works.
<!-- Reviewable comment -K295MMeoO7GldBbssem:-2146843840 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 136 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/287#-K294gMdxrbzjWZ2UHu3)** ([raw file](https://github.com/codice/ddf/blob/d5dce05fe7a86507d21a80e645420d3ad50185af/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L136)):</sup>
`AbstractIntegrationTest` is getting on the long side. Should we consider extracting it to it's own file?

We should at least add JavaDoc to explain what it is and how to use it.
<!-- Reviewable comment -K294gMes5ycZpQefiUy:1983721772 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

harrison-tarr commented:
Review status: 6 of 7 files reviewed at latest revision, 21 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java, line 136 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/287#-K294gMdxrbzjWZ2UHu3)** ([raw file](https://github.com/codice/ddf/blob/d5dce05fe7a86507d21a80e645420d3ad50185af/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/AbstractIntegrationTest.java#L136)):</sup>
Added Javadoc
<!-- Reviewable comment -K29R6vnsR26b5eJRAvo:-1820831855 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

AzGoalie commented:
Reviewed 1 of 6 files at r3, 1 of 3 files at r6, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java, line 92 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/287#-K1f00rpcjHRkHdQtsSO)** ([raw file](https://github.com/codice/ddf/blob/7f4a0519c6b74692c0651fb871ba7b6b7f67ea21/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestCatalog.java#L92)):</sup>
Should add documentation so we don't forget about this.
<!-- Reviewable comment -K2DOeVawiIlaUpYAKu-:1943215701 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

AzGoalie commented:
LGTM!
![selfie-0](http://i.imgur.com/mQDn2KP.png)

lessarderic commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Ship it!

coyotesqrl commented:
I believe it is time to squash commits and light this candle. :+1:

ricklarsen commented:
Reviewed 1 of 6 files at r3, 1 of 3 files at r5, 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/287)
<!-- Sent from Reviewable.io -->

ricklarsen commented:
built & tests passed & verified 2 concurrent builds.
